csharp/src/Ice/OutgoingResponseFrame.cs

```diff
@@ -105,53 +100,94 @@ public OutgoingResponseFrame(Current current, RemoteException exception)
                         Debug.Assert(false);
                     }
 
-                    WriteByte((byte)replyStatus);
-                    dispatchException.Id.IceWrite(this);
+                    ostr.WriteByte((byte)replyStatus);
+                    dispatchException.Id.IceWrite(ostr);
 
                     // For compatibility with the old FacetPath.
                     if (dispatchException.Facet == null || dispatchException.Facet.Length == 0)
                     {
-                        WriteStringSeq(Array.Empty<string>());
+                        ostr.WriteStringSeq(Array.Empty<string>());
                     }
                     else
                     {
-                        WriteStringSeq(new string[] { dispatchException.Facet });
+                        ostr.WriteStringSeq(new string[] { dispatchException.Facet });
                     }
-                    WriteString(dispatchException.Operation);
+                    ostr.WriteString(dispatchException.Operation);
 
                 }
                 else
                 {
-                    WriteByte((byte)ReplyStatus.UnknownLocalException);
+                    ostr.WriteByte((byte)ReplyStatus.UnknownLocalException);
                     if (requestFailedException.IceMessage.Length > 0)
                     {
-                        WriteString(requestFailedException.IceMessage);
+                        ostr.WriteString(requestFailedException.IceMessage);
                     }
                     else
                     {
-                        WriteString(requestFailedException.ToString());
+                        ostr.WriteString(requestFailedException.ToString());
                     }
                 }
             }
             else
             {
-                WriteByte((byte)ReplyStatus.UserException);
-                StartEncapsulation(current.Encoding, FormatType.SlicedFormat);
-                WriteException(exception);
-                EndEncapsulation();
+                ostr.WriteByte((byte)ReplyStatus.UserException);
+                ostr.StartEncapsulation(Encoding, FormatType.SlicedFormat);
+                ostr.WriteException(exception);
+                ostr.EndEncapsulation();
             }
+            OutputStream.Position payloadEnd = ostr.Tail;
+            ostr.Finish();
```

Comment:
`Finish` could return the tail instead here this would be coherent with the OutputStream constructor which is given `PayloadBegin`. And Tail could remain private to OutputStream.